src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodTranslator.cs

```diff
@@ -55,12 +55,14 @@ public SqlExpression Translate(SqlExpression instance, MethodInfo method, IReadO
             {
                 return
                     _sqlExpressionFactory.GreaterThan(
-                        _jsonTranslator.TranslateArrayLength(arrayOperand) ??
-                        _sqlExpressionFactory.Function(
-                            "array_length",
-                            new[] { arrayOperand, _sqlExpressionFactory.Constant(1) },
-                            typeof(int)),
-                        _sqlExpressionFactory.Constant(1));
+                        _sqlExpressionFactory.Coalesce(
```

Comment:
First, there's no need to coalesce the JSON version of array length, since `json_array_length` returns 0 and not null for empty arrays (`SELECT json_array_length('[]')`). So you can move the coalesce inside.

Interesting theoretical note: for the usual case with Any, we don't have to coalesce even for array_length, because `SELECT * WHERE array_length(arr) > 0` will still work as expected: `NULL > 0` returns NULL which evaluates to false in a predicate. However, if the expression is negated we get wrong results: `SELECT * WHERE NOT array_length(arr) > 0`, since negated NULL is still NULL. So we should coalesce here as well.

Reply:
There is a problem. `json_array_length` behaves differently and returns 0 for an empty array while `array_length` returns `NULL`.

```sql
SELECT
    array_length(NULL::int[], 1),
    array_length(ARRAY[]::int[], 1),
    array_length(ARRAY[1,2,3], 1),
    json_array_length(NULL),
    json_array_length('[]'),
    json_array_length('[1,2,3]');
```

```
 array_length | array_length | array_length | json_array_length | json_array_length | json_array_length
--------------+--------------+--------------+-------------------+-------------------+-------------------
              |              |            3 |                   |                 0 |                 3
```

But if I change the code as you suggested, results will look like this:

```sql
SELECT
    COALESCE(array_length(NULL::int[], 1), 0),
    COALESCE(array_length(ARRAY[]::int[], 1), 0),
    COALESCE(array_length(ARRAY[1,2,3], 1), 0),
    json_array_length(NULL),
    json_array_length('[]'),
    json_array_length('[1,2,3]');
```

```
 coalesce | coalesce | coalesce | json_array_length | json_array_length | json_array_length
----------+----------+----------+-------------------+-------------------+-------------------
        0 |        0 |        3 |                   |                 0 |                 3
```

As you can see there is a difference again, so `COALESCE` for the both methods unifies behavior. It's acceptable as you said, but in some cases it won't work (for example, when result of `Any` is compared with some variable).

```sql
SELECT
    (COALESCE(array_length(NULL::int[], 1), 0) > 0) = false,
    (COALESCE(array_length(ARRAY[]::int[], 1), 0) > 0) = false,
    (COALESCE(array_length(ARRAY[1,2,3], 1), 0) > 0) = false,
    (json_array_length(NULL) > 0) = false,
    (json_array_length('[]') > 0) = false,
    (json_array_length('[1,2,3]') > 0) = false;
```

```
 ?column? | ?column? | ?column? | ?column? | ?column? | ?column?
----------+----------+----------+----------+----------+----------
 t        | t        | f        |          | t        | f
```

Reply:
Understood - so we'd add coalesce on json_array_length for the case where the array itself is null. I'm OK with that.

Note that there are many other cases (not just arrays) where things probably don't behave 100% correctly around nulls. These cases - which in LINQ to Objects would throw an NRE - are pretty hard to get 100% right. That's why I'm much more concerned with empty arrays than with null arrays.